### PR TITLE
Add TraitAssembler

### PR DESCRIPTION
**Pull request by malles**
Adds a Trait and Use statement to a TypeContext class.

### CODE REVIEW

File: src/Phpro/SoapClient/CodeGenerator/Assembler/TraitAssembler.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * Class TraitAssembler
+ *
+ * @package Phpro\SoapClient\CodeGenerator\Assembler
+ */
+class TraitAssembler implements AssemblerInterface
+{
+    /**
+     * @var
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/TraitAssembler.php]: The type of the var is missing in the docblock

File: src/Phpro/SoapClient/CodeGenerator/Assembler/TraitAssembler.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    /**
+     * @param ContextInterface|TypeContext $context
+     */
+    public function assemble(ContextInterface $context)
+    {
+        $class = $context->getClass();
+
+        try {
+            if (!in_array($this->traitName, $class->getUses())) {
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/TraitAssembler.php]: If you add a use with an alias, you can't use in_array like this.
The uses are stored like this:

`MyNamespace\Something\Something as MyAlias`

It is also possible that the uses contain a trailing backslash. I guess this logic needs a bit more work.

File: src/Phpro/SoapClient/CodeGenerator/Assembler/TraitAssembler.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                $class->addUse($this->traitName, $this->traitAlias);
+            }
+			$traitAlias = $this->traitAlias;
+			if (!$traitAlias) {
+				$a = explode('\\', $this->traitName);
+				$traitAlias = array_pop($a);
+			}
+			$traits = $class->getTraits();
+			if (!in_array($traitAlias, $traits)) {
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/TraitAssembler.php]: I guess this part works the same as the getUses part. It might require some normalization.

File: src/Phpro/SoapClient/CodeGenerator/Assembler/TraitAssembler.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    public function assemble(ContextInterface $context)
+    {
+        $class = $context->getClass();
+
+        try {
+            if (!in_array($this->traitName, $class->getUses())) {
+                $class->addUse($this->traitName, $this->traitAlias);
+            }
+			$traitAlias = $this->traitAlias;
+			if (!$traitAlias) {
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/TraitAssembler.php]: Maybe this logic can be moved to a Util class so that it canbe reused?

**Comment by veewee** [comment]
Thanks for the PR!
Can you resolve the Travis issues? It looks like you've used tabs instead of 4 spaces.
Before I can merge this PR, we'll also need a phpunit test which also covers the edge cases as commented in the source code. We also need some documentation about the generator so that it can be used by everybody.

**Comment by veewee** [comment]
I was also thinking: maybe it's a good idea to split the use part into a separate `UseAssembler` which can be reused in the other classes.

**Comment by malles** [comment]
I fixed the style issues. I'll check the logic and separation of the UseTrait later.
Are you on [Gitter](https://gitter.im/malles)? That might be easy to discuss some ideas.

**Comment by veewee** [comment]
I've created a room: https://gitter.im/phpro/soap-client
You might also find me as 'VeeWee' on Freenode.

File: src/Phpro/SoapClient/CodeGenerator/Assembler/TraitAssembler.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        $class = $context->getClass();
+
+        try {
+            $useAssembler = new UseAssembler($this->traitName, $this->traitAlias);
+            if ($useAssembler->canAssemble($context)) {
+                $useAssembler->assemble($context);
+            }
+
+            $traitAlias = $this->traitAlias;
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/TraitAssembler.php]: Maybe it's shorter / more readable to just:
```php
$traitAlias = $this->traitAlias ?: Normalizer::lastPart($this->traitName);
```

File: src/Phpro/SoapClient/CodeGenerator/Assembler/UseAssembler.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        $this->useAlias = $useAlias;
+    }
+
+    /**
+     * @param ContextInterface $context
+     * @return bool
+     */
+    public function canAssemble(ContextInterface $context)
+    {
+        return $context instanceof TypeContext;
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/UseAssembler.php]: This means that a PropertyContext will never be able to add a use? 
Maybe in some situations it might be handy  to be able to add a Use in a method?

File: src/Phpro/SoapClient/CodeGenerator/Assembler/UseAssembler.php [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            }
+        } catch (\Exception $e) {
+            throw AssemblerException::fromException($e);
+        }
+    }
+
+    /**
+     * @return string
+     */
+    protected function fullName()
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/UseAssembler.php]: Also this method I would like to see in the util class. Maybe something like `getCompleteUseStatement($useName, $useAlias)`?

File: src/Phpro/SoapClient/CodeGenerator/Assembler/UseAssembler.php [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * @param ContextInterface|TypeContext $context
+     */
+    public function assemble(ContextInterface $context)
+    {
+        $class = $context->getClass();
+
+        try {
+            if (!in_array($this->fullName(), $class->getUses())) {
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Assembler/UseAssembler.php]: What about classes that are already imported but without the specified alias?

File: src/Phpro/SoapClient/CodeGenerator/Util/Normalizer.php [hunk 9]
```diff
@@ -66,4 +66,14 @@ public static function generatePropertyMethod($prefix, $property)
     {
         return strtolower($prefix) . ucfirst(self::normalizeProperty($property));
     }
+
+    /**
+     * @param string $name
+     * @return string
+     */
+    public static function lastPart($name)
```
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Util/Normalizer.php]: Maybe `lastPart()`  isn't such a good name?
What do you think about something like `getClassNameFromFQN`?

File: test/PhproTest/SoapClient/Unit/CodeGenerator/Assembler/UseAssemblerTest.php [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+{
+
+
+}
+
+CODE;
+
+        $this->assertEquals($expected, $code);
+    }
+
```
Comment (veewee) [inline test/PhproTest/SoapClient/Unit/CodeGenerator/Assembler/UseAssemblerTest.php]: I would like to add some additional tests for this class:

- Use with an alias.
- Existing use without alias

File: test/PhproTest/SoapClient/Unit/CodeGenerator/Assembler/TraitAssemblerTest.php [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    use MyTrait;
+
+
+}
+
+CODE;
+
+        $this->assertEquals($expected, $code);
+    }
+
```
Comment (veewee) [inline test/PhproTest/SoapClient/Unit/CodeGenerator/Assembler/TraitAssemblerTest.php]: We'll need an additional test with an aliasName set.

File: src/Phpro/SoapClient/CodeGenerator/Util/Normalizer.php (hunk 9 above)
Comment (veewee) [inline src/Phpro/SoapClient/CodeGenerator/Util/Normalizer.php]: This class is tested with phpspec. You can find the tests in the spec folder.
Also for this class we'll need to add tests.

**Comment by veewee** [comment]
Hi @malles,

I've added some remarks to the PR. It's starting to look pretty nice. Looking forward to merge this one in :)
Great job!

File: docs/code-generation/assemblers.md [hunk 12]
```diff
@@ -16,6 +16,8 @@ to generate the code you want to add to the generated SOAP types.
 - [ResultProviderAssembler](#resultproviderassembler)
 - [SetterAssembler](#setterassembler)
 - [FluentSetterAssembler](#fluentsetterassembler)
+- [TraitAssembler](#traitassembler)
```
Comment (veewee) [inline docs/code-generation/assemblers.md]: The items were ordered alfabetically. Can you remain this sort order?

**Comment by veewee** [comment]
@malles Are you planning to finish this PR any time soon? Thanks!

**Comment by veewee** [comment]
@malles, Can you enable this option so that I can easily finish this PR for you?
https://github.com/blog/2247-improving-collaboration-with-forks

I am going to write some additional generators which are depending on this changes.
Thanks!

**Comment by veewee** [comment]
Thanks for your work @malles. I wanted to move on with my PR so I finished up yours.
